build/do.rq

```diff
@@ -163,6 +184,131 @@ write_readme {
+	#		"tags",
+	#		"v1.0.1"
+	#	],
+	#	[
+	#		"bab5bdd4bb2de45c7d1bcf5d7b5df3c61a131f38\trefs",
+	#		"tags",
+	#		"v1.1.0"
+	#	],
+	#	...
+
+	print("fetching tags for enterprise-opa repository")
+
+	eopa_tags_result := rq.run([
+			"git",
+			"ls-remote",
+			"--tags",
+			"https://github.com/styrainc/enterprise-opa"
+		], {
+			"stdout_spec": {
+				"format": "raw",
+				"options": {
+					"raw.fs": "/",
+					"raw.rs": "[\n\r]"
+				},
+			}
+		}
+	)
+
+	{ rq.error(sprintf("\nstdout: %s\nstderr: %s\n", [eopa_tags_result.stdout, eopa_tags_result.stderr])) | eopa_tags_result.exitcode != 0 }
+
+
+	# We assume that tags and capabilities files are 1:1, but some EOPA
+	# release tags in the past did not correctly get capabilities files, so
+	# we eliminate them from consideration.
+	
+	known_bad_tags := {
+		"v1.15.0",   # tag missing capabilities file (misnamed v0.15.0)
+		"v1.4.1",    # tag missing capabilities file
+		"v1.5.0",    # tag missing capabilities file
+	}
+
+	# Note that we use the `not startswith` to explicitly drop any
+	# pre-1.0.0 tags. There is little reason anyone would want them, and if
+	# they really have a need, they can manually download them and use the
+	# file locally.
+
+	eopa_tags := {t | some r in eopa_tags_result.stdout; t := r[2] ; not known_bad_tags[t] ; not startswith(t, "v0.")}
+
+	# Get a directory listing for the capabilities directory, filtering for
+	# only nonzero size files with JSON extensions. The size check is to
+	# avoid long-tail edge cases where we crashed after opening the file
+	# for writing but before committing any content.
+	eopa_caps_tree := {p: f | f := rq.tree(eopa_caps_dir, {})[p]; f.size != 0; f.ext == "json"}
+
+	# Determine which capabilities files are missing, what URL they
+	# should be fetched from, and where they should end up on disk.
+	missing_locally := {
+		{"local": p, "remote": r}
+
+		|
+
+		# construct the local path we expect the caps to exist at
+		t := eopa_tags[_]
+		p := rq.joinpath([eopa_caps_dir, sprintf("%s.json", [t])])
+
+		# eopa_caps_tree has object values, so if the path is missing,
+		# retrieving a non-object type is sufficient to check for
+		# nonexistance
+		object.get(eopa_caps_tree, p, "-") == "-"
+
+		# construct the URL to fetch the content from
+		r := rq.template("https://raw.githubusercontent.com/StyraInc/enterprise-opa/main/capabilities/{{.tag}}.json", {"tag": t})
+	}
+
+	print(sprintf("fetching %d capabilities files missing locally", [count(missing_locally)]))
+
+	# Download the capabilities from the constructed URLs.
+	new_caps := {
+		{"local": m.local, "content": c}
+		|
+		m := missing_locally[_]
+		print("\tfetching ", m.remote)
+		resp := http.send({"url": m.remote, "method": "GET"})
+			
+		{ rq.error(sprintf("non-200 status code '%d' for URL '%s'", [resp.status_code, m.remote])) | resp.status_code != 200 }
+
+		c := resp.raw_body
+
+		# sanity check in case we got the URL wrong
+		not regex.match("404: Not Found", c)
```

Comment:
this would have failed on `resp.status_code != 200 ` already, no?

Reply:
b5cb5c5